### PR TITLE
Interactive SpyQL

### PR DESCRIPTION
**Pull request by yashbonde**
This is v0 for interactive spyql. It implements the following new things:

1. `Q` class and `q` function in `interactive.py`: This is the class/function that the user can add in their script
2.  `InteractiveProcessor` in `processor.py`: row yielding loaded by the `Processor`
3.  `InteractiveWriter` in `processor.py`: writer that adds the output in a list

Other Important changes:
1. `init_vars`, `go` and `_go` now have `user_query_vars` as a valid input
2. `Processor` can now be explicitly defined as interactive

Misc:
1.  `if`s changed to `elif`s where needed

Issue: #64

To test this `python tests/interactive_test.py`.

### CODE REVIEW

**Comment by dcmoura** [comment]
Thank you so much for contributing @yashbonde !

I think what you wrote is great. I have several suggestions that I will include when reviewing the PR but in the meanwhile I would like you to think in the following things we might want to do:
- allow querying files and not just data structures. e.g. we might want to read a very large file, do some aggregations that would result in a smaller output, and return the result (otherwise, we would need to load the large file into an in-memory datastructure)
- allow writing the result to a file, and not necessarily limit the output to a data structure (I haven’t a strong option about this one)
- support multiple data structures types on input: list of dicts, pandas, …, or at least offer ways to convert commonly used data structure to a standard one that spyql can handle
- have different output data structure types or return an object with functions to convert to other formats
- have a writer which is a lambda function, letting to the lambda function deal with the outputs rows, one by one

I am not saying that a first approach to interactive spyql should cover all these, but at least we should be mindful of decisions we make about these and eventual other relevant topics. Pandas is super popular but I have been avoiding it due to the large number of dependencies... 

Let me know what you think!

**Comment by yashbonde** [comment]
Hi,

Your points make sense. I would go about it this way:

1. Supporting multiple sources in the interactive form: This has to be done by expanding the grammar a little bit:
  a. when using a python data structure: `FROM xxx` becomes `FROM {xxx}` (not the curly braces). We will know immediately what are the python inputs expected at runtime from the user. This will set `prs["from"] = "PYTHON"`
  b. we can add checks in the `Q.__call__` that see what datatype the object is, this will determine the `Processor` to use, ex. `PandasProcessor, PolarsProcessor`
  c. `FROM` and `TO` can be written down as (relative) filepaths: `FROM ./data.csv TO ./data.json`. This will set `prs["from"] = "JSON/CSV/TEXT"` and can be processed as it currently being processed. Difference being that the if `TO` is not provided then the information is stored to the memory and returned as python data structure (sounds like a fancy file reader ;-) )
2. Returning a new structure (`SpyOutput`) can be a cool strategy of further exporting it wherever the user wants. It can have function to send it anywhere `out.to_csv("filepath")` or `out.to_list()`, etc. I think I'll work on it once the core logic mentioned above is working.
3. Writer as a lambda? I don't get it, but pure functional approach is something I am happy to work with.

Things I'll implement:
→ I am trying to get all the tests in `main_test.py` working interactively
→ Pandas and Polars in working condition

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:69]:
```suggestion
```
I might me overlooking something, but it seems to me that we do not need an `InteractiveProcessor`. The  `PythonExprProcessor` would do just fine since, with the change I suggest in the `init_vars` method, all variables are directly available in the scope of the spyql query.

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:75]:
```suggestion

```

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:88]: I would revert  this change.
`elif`s are just a matter of style here since if the previous `if` holds true, then the method returns. So the `else` is not doing much... I prefer using `if`s in this case, but not sure if there are any styling guidelines about this. Maybe aiming for readability would be best...

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:95]:
```suggestion
        spyql.log.user_dbg(f"Loading {self.__class__.__name__}")
        self.prs = prs  # parsed query
        spyql.log.user_dbg(self.prs)
```
I prefer leaving this kind of internal outputs to debug level... I believe `prs` is already being outputted in the CLI, so moving to here might make sense.

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:97]:
```suggestion

```

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:287]: I think if we revert this change everything should work as expected.
Regarding the `data` alias, we can add it as a translation in the constructor of the`PythonExprProcessor`, e.g.

   `self.translations["data"] = "_values"`

This will result in every reference to `data` being replaced by a reference to `_values`.

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:327]: This is nice. 
One note is that if we decide to return a result object in the future, we might also include statistics like the #rows in/out, among other potential helpful stuff

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:454]: I think we do not need this class, and it's actual very redundant vs `PythonExprProcessor`. 
But please assess.

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:484]: I would suggest Processors to receive a `input_file` rather than a path.  This file would default to `sys.stdin`. 
This way we would move the boilerplate of file opening to a single place.
Then, reading would simply be:
`return ([line.rstrip("\n\r")] for line in input_file)`

This applies to all processors.

File: spyql/processor.py
Comment (dcmoura) [inline spyql/processor.py:482]: I believe this is equivalent to using a generator like this (but never compared performance):
`([line.rstrip("\n\r")] for line in f)`

File: spyql/utils.py
Comment (dcmoura) [inline spyql/utils.py:38]:
```suggestion
def get_folder(x):
```
suggesting using a verb to illustrate what function does

File: spyql/utils.py
Comment (dcmoura) [inline spyql/utils.py:42]:
```suggestion
def join_paths(x, *args):
```
some day we might have queries with `JOIN`s, and this would cause some confusion :-)

File: spyql/writer.py
Comment (dcmoura) [inline spyql/writer.py:21]: again, `elif` is just a matter of style, but if you think it makes it more readable, go ahead

File: spyql/writer.py
Comment (dcmoura) [inline spyql/writer.py:60]: What about using or extending the class `CollectWriter`?
https://github.com/dcmoura/spyql/blob/c379afa89c883b29bc4d3eaf3eea5a7c9d9cbd91/spyql/writer.py#L94

File: tests/interactive_test.py
Comment (dcmoura) [inline tests/interactive_test.py:2]: One thing we might do is adding an interactive testing here:
https://github.com/dcmoura/spyql/blob/c379afa89c883b29bc4d3eaf3eea5a7c9d9cbd91/tests/main_test.py#L67

This way, all tests done to the CLI would be also done in interactive mode. Then, we would only need to tackle here features that are specific to the interactive mode.

I will skip this file for this review and come back in the next one.

File: tests/main_test.py
Comment (dcmoura) [inline tests/main_test.py:89]: Didn't get this... probably you needed to turn off testing temporarily?

**Review by dcmoura — CHANGES_REQUESTED** [review]
Great work!
I am very impressed, you got a good grasp of the codebase rather quickly. 

I have several suggestions that I would like you to go through. 
Some general remarks:
- we should try to make the CLI and the interactive mode as similar as possible, this might mean moving some logic that you added to a place where it is shared with the CLI
- IMO, we do not need to clearly state the mode we are working on: we are either using this in as a CLI or as a Python lib, and this should be irrelevant to most components of spyql, mainly to the processor (I have addressed this in the comments)
- we should think on how to address Input/Output options of the Processor/Writer, I might be missing something but it seems that this is not being addressed? And also other options that we would pass in the command-line in general...

Thank you so much, hope you are having fun! :-)

**Comment by dcmoura** [comment]
> 3. Writer as a lambda? I don't get it, but pure functional approach is something I am happy to work with.

Vary basic (actually silly) example:
`SELECT col1 FROM range(10) TO lambda(row): print(row)`

This would allow users to have custom writers very easily... without having to extend the writer and register it (not possible at the moment). But I guess this is for a future iteration if it makes sense.

File: spyql/interactive.py
Comment (yashbonde) [reply spyql/interactive.py:70]: The values here are more about the structure of the output, whereas in this case it is telling more about the output location (bad word?), which is the caller. But I get the idea that user can get output in multiple format, I think I'll implement it in the `Output` object which is returned by the `go`. The `Output` can have functions like `to("dict")` or dedicated methods `to_list, to_records, ...`.

When I work on the output I'll implement as `to()`, will change this before final merge.

File: spyql/processor.py
Comment (yashbonde) [reply spyql/processor.py:87]: I read the Google's [styleguide](https://google.github.io/styleguide/pyguide.html), there's nothing mentioned. Anyways I prefer using `elif`s because when the code AST is generated the entire conditional block looks like a single operation, whereas in case of bunch of `if`s, each step is treated like a different operations.

Screenshot 2022-02-20 at 12 05 22 PM

vs.

Screenshot 2022-02-20 at 12 05 51 PM

But I'll leave it to you, whatever you prefer, for now I am leaving it as elifs.

File: spyql/writer.py
Comment (yashbonde) [reply spyql/writer.py:60]: Yeah, I missed this. I am removing the Interactive writer and using CollectWriter.

File: spyql/processor.py
Comment (yashbonde) [reply spyql/processor.py:481]: Cannot do this because it throws the error:
```
  File "/Users/yashbonde/Desktop/ai/repos/spyql/spyql/processor.py", line 525, in <genexpr>
    return ([decoder.decode(line)] for line in f)
ValueError: I/O operation on closed file.
```

File: spyql/processor.py
Comment (yashbonde) [reply spyql/processor.py]: Removed

File: spyql/processor.py
Comment (yashbonde) [reply spyql/processor.py:67]: I think there are couple of options:
1. Reserve a few words
2. Warn user that there is an override of default functions and a list of them down (I like this)

File: spyql/processor.py
Comment (yashbonde) [reply spyql/processor.py]: Haha, one who creates it definitely knows their code better 😂. Struggled hard on this one, adding `self.translations["data"] = "_values"` in `PythonExprProcessor.__init__` works! 🙏

File: spyql/interactive.py
Comment (yashbonde) [reply spyql/interactive.py]: I have added the check in `cli.run`, since the requirements by CLI and Interactive can differ (python objects can be passed in Interactive) adding the logic in parse did not make sense.

**Comment by yashbonde** [comment]
> CLI tests are still failing and I don't know why

This was fun, I think we can make some serious progress in this direction if we are able to support:
1. Purely functional approach so multi-thread (Interactive and CLI) and multi-core (CLI only) combined can be used reliably (see `vmap/pmap` in `jax`). This is completely possible but will require a big refactor to make it completely functional
2. Figure out how to use multiple files and "JOIN" them.
3. Support wider I/O with the package in terms of:
  * Auto replace larger operations with `numpy/jax` arrays to utilise more compute
  * dataframes and rely on their internals for many of the operations, will require writing a new `DataFrame` class that can have a `make_dataframe` that takes in the string and tries to get it actual Dataframe object.
  * Networking support ex. `SELECT * FROM aws.s3://object` can automatically download/stream the files
  * Connecting with an actual DB (like Sqlite3 or Postgres) ?? (still unsure about this)

Personally I think these are the places where it can be super useful:
1. Default CRUD app logic can be made really simple by chaining a bunch of spyql queries

File: spyql/interactive.py
Comment (dcmoura) [reply spyql/interactive.py:71]: `"MEMORY"` makes sense! For context, spyql once had a `"PY"` format that was similar to the `"SPY"` format but in plain text. 
Yes, let's go with the `to` approach.
What do you think about leaving the result object implementation as well as support  to follow-up PRs? In this PR we would try to set the foundation for interactive spyql and then we would add new features. The idea is simply to don't make this PR too long, but if you prefer we can continue iterating on this one.

File: spyql/processor.py
Comment (dcmoura) [reply spyql/processor.py]: Makes sense leaving as `elif`s since we are emulating a `switch ... case`, which would be a single expression.

File: spyql/processor.py
Comment (dcmoura) [reply spyql/processor.py]: This would have to happen inside the `with` block... but again, seems more a matter of style and having more compact code. I will revisit this when I review your latest changes

**Comment by dcmoura** [comment]
> > CLI tests are still failing and I don't know why

I will check what's happening, might be some dirty detail :-)

> I really want this to be used more widely, but the fundamental problem is that Python is inherently slow. So if I have 2GB of data that I have to query again and again it would make sense to load that into a DB. 

True. spyql is great if your analysis fits into a single query (or a chain of queries). If you need to repeat different queries over the same (large/big) data, spyql might not be the best solution. In that case, depending on the size of the data, I would jump to either PostgreSQL (and spyql helps here allowing to easily import data into the DB using the `TO sql`) or Spark-SQL (typically in standalone mode after an initial conversion of the data to parquet).

> The strong spot of spyql is:
> 
> 1. When dataset size is low (need benchmarking and numbers on this)
> 2. When querying is a one time process.
> 

True :-) If the dataset is small or if don't need to do multiple different queries on the same large data.
Here's a simple benchmark for json data. spyql is head to head to `jq`. 

spyql_json_simple_benchmark

More info in my FOSDEM22 presentation:
- [Slides](https://fosdem.org/2022/schedule/event/python_spyql/attachments/slides/5054/export[…]/attachments/python_spyql/slides/5054/spyql_fosdem22_slides.pdf)
- [Presentation recording](https://vimeo.com/665833922) 

> Personally I think these are the places where it can be super useful:
> 
> 1. Default CRUD app logic can be made really simple by chaining a bunch of spyql queries
> 2. Please mention if you can come up with more!
> 

I think that CLI can leverage the command-line much better than python alone for data processing. In this context, spyql is great for:
- Ad-hoc querying / data analyses
- Converting file formats (e.g. CSV, JSON) and importing data to DBs
- Setup simple data pipes
- Monitoring data streams
- Transforming/analysing outputs from shell commands


> This was fun, 
Yes!! :-D

> I think we can make some serious progress in this direction if we are able to support:
Great points you have here.

> 
> 1. Purely functional approach so multi-thread (Interactive and CLI) and multi-core (CLI only) combined can be used reliably (see `vmap/pmap` in `jax`). This is completely possible but will require a big refactor to make it completely functional

This seems exciting. My main goal now is making spyql feature-rich while having a decent performance. I am trying to understand what could lead to higher adoption rates. Most of the feedback I am getting is about features for making query writing easier like: auto-complete (including column names) and syntax highlighting. So, I am planning to tackle IPython integration, and the work you are doing put us already half-way there. See for instance: https://github.com/catherinedevlin/ipython-sql

> 2. Figure out how to use multiple files and "JOIN" them.

For now I would keep this very simple. Load a JSON into memory (as a dict) and then do `SELECT my_json[my_foreighn key] FROM ...`. Basically, this allows for equi-joins using dict lookups. 


> 3. Support wider I/O with the package in terms of:
> 
> * Auto replace larger operations with `numpy/jax` arrays to utilise compute more effectively

This seems a nice entry for the backlog, given that I don't think performance is an adoption barrier right now.


> * dataframes and rely on their internals for many of the operations, will require writing a new `DataFrame` class that can have a `make_dataframe` that takes in the string and tries to get it actual Dataframe object.

Interesting

> * Networking support ex. `SELECT * FROM aws.s3://object` can automatically download/stream the files

This is already possible with the CLI using the ASW cli, and piping stdout to spyql. But I agree that this will make it much easier. 

> * Connecting with an actual DB (like Sqlite3 or Postgres) ?? (still unsure about this)

Also possible with the CLI by piping the output of DB clients, but again, handy to be included for interactive use.

> * Storage of information in some highly optimised format such as Apache Parquet can fix the issues with large dataset size and `mmap` of `pyarrow` and `numpy` and `numpy` to `jax` can enable **zero-copy**, **high-performance** processing.

This is definitely something to consider on the mid-term.


> If we go by the logic that we keep spyql as the Front end and build plugins that can be used with other systems (pull from S3 process data on GPUs and return a dict object).

This is very interesting... definitely something to explore, especially if anyone could contribute with plugins!

> Note sure if these pointers are useful, but I learnt a lot hacking this. It's super cool 👍

Super useful and insightful! I am super happy with both what you are doing and also hearing your perspective! 
Thank you!!

File: spyql/interactive.py
Comment (yashbonde) [reply spyql/interactive.py:71]: Sure. I think for this version you can keep the `ProcessedOutput` object since there is no change in CLI behaviour except a minor `spyql.log.user_info(f"Output Meta: {out}")` added at the end of `main()`.

File: spyql/processor.py
Comment (yashbonde) [reply spyql/processor.py]: I did implement it in a `with` like follows but threw the error as above, maybe we can leave it as is for now:

```
with open(...):
  return (...)
```

**Comment by yashbonde** [comment]
@dcmoura go for it! I have pushed all my changes in `yashbonde:master`. You can merge it into a feature branch as it would make more sense.

* Piping through CLI for AWS etc. makes way more sense than adding them here unless someone really requests it.

* I was not aware of the Performance metrics but they are very good and your talk demonstrated far more than I had in mind. I work with data pretty rarely, mostly it's building a new a [ML pipeline](https://github.com/NimbleBoxAI/nbox/) (apologies for the shameless plug 😛) and so I was a bit concerned about the metrics as I don't have experience in the domain.

I read a bit on [LINQ](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/) and it is a very fascinating idea. Ability to effectively query complicated data structure from within a language (even if using another language which is a standard) can reduce the amount of time and code required.

> ... � So, I am planning to tackle IPython integration, and the work you are doing put us already half-way there.

I actually was looking for something in notebooks then landed back here. I may have put you half way there but you came up with the whole thing so you deserve far more credits. I'll take a look at notebook thing as well!

Best,
Yash

**Comment by dcmoura** [comment]
Thank you @yashbonde :-) 

LINQ in C# is really cool, as it seems to be handled at compile-time (and not a query in a string evaluated at runtime).
I also noticed how they changed the order of the clauses to match the execution order (an option that I am also wondering if should be implemented).

About Jupyter plugins for SQL (and actually much more) you can also take a look to this other project: https://github.com/fugue-project/fugue

I'll ping you when I am done or if I have any question!
Thanks,
Daniel

**Comment by dcmoura** [comment]
Hi @yashbonde . Yesterday I have managed to finishing fixing the code and do some very much needed refactoring to eliminate code/logic replication. The code was breaking on lots of places and took me some hours to fix everything (I had to revert several changes). 
Next time, please interrupt coding when you detect that tests are not passing. Every time I change something I run `make test` to quickly check if all tests are passing and run `black` to format the code before committing. 

Your contribution was important and timely but please reach out to me when tests start breaking next time, OK? 

I am taking the opportunity to make further changes/refactorings, so I will take a bit longer to submit the PR, but if you are interested you can check the changes I have made here: https://github.com/dcmoura/spyql/compare/interactive

Thanks

**Comment by yashbonde** [comment]
Got it, will take care of it from next time!

**Comment by dcmoura** [comment]
Hi @yashbonde , the PR is finally out! 
https://github.com/dcmoura/spyql/pull/68

Feel free to provide feedback. Thank you for your help!

**Pull request merged**
